great_expectations/data_context/store/datasource_store.py

```diff
@@ -172,22 +167,31 @@ def set_by_name(
         )
         self.set(datasource_key, datasource_config)
 
-    def create(
-        self, datasource_config: DatasourceConfig
-    ) -> Optional[GeCloudResourceRef]:
+    def set(  # type: ignore[override]
+        self, key: Union[DataContextKey, None], value: DatasourceConfig, **_: dict
+    ) -> DatasourceConfig:
         """Create a datasource config in the store using a store_backend-specific key.
-
         Args:
-            datasource_config: Config containing the datasource name.
-
+            key: Optional key to use when setting value.
+            value: DatasourceConfig set in the store at the key provided or created from the DatasourceConfig attributes.
+            **_: kwargs will be ignored but accepted to align with the parent class.
         Returns:
-            None unless using GeCloudStoreBackend and if so the GeCloudResourceRef which contains the id
-            which was used to create the config in the backend.
+            DatasourceConfig retrieved from the DatasourceStore.
         """
-        key: Union[
-            GeCloudIdentifier, DataContextVariableKey
-        ] = self._build_key_from_config(datasource_config)
-        return self.set(key, datasource_config)  # type: ignore[func-returns-value]
+        if not key:
+            key = self._build_key_from_config(value)
+
+        # Make two separate requests to set and get in order to obtain any additional
+        # values that may have been added to the config by the StoreBackend (i.e. object ids)
+        ref: Optional[Union[bool, GeCloudResourceRef]] = super().set(key, value)  # type: ignore[func-returns-value]
```

Comment:
... does the return type have to be optional?

Reply:
It does, currently at least the `InMemoryStoreBackend` and `InlineStoreBackend` return None on `_set()` which propagates up to this return type. We can standardize and return a key for all backends but I'd prefer to do that (and remove the bool type) in a separate effort/PR in case it is not trivial.